File: 2015/day04.py
import fileinput
import hashlib


def first_with_prefix(secret, prefix):
    i = 1
    root_hash = hashlib.md5(secret.encode())
    while True:
        h = root_hash.copy()
        h.update(str(i).encode())
        if h.hexdigest().startswith(prefix):
            return i
        i += 1


def part1(secret):
    return first_with_prefix(secret, '00000')


def part2(secret):
    return first_with_prefix(secret, '000000')


if __name__ == "__main__":
    secret = next(fileinput.input()).strip()

    print("Part 1: %s" % part1(secret))
    print("Part 2: %s" % part2(secret))

File: 2015/rust/src/bin/day04.rs
use anyhow::Result;
use md5::{Digest, Md5};

use aoc::utils::get_input;

fn main() -> Result<()> {
    let input = get_input()?;
    let secret = input.trim();

    println!("Part 1: {}", part1(secret));
    println!("Part 2: {}", part2(secret));

    Ok(())
}

fn part1(secret: &str) -> usize {
    find_hash(secret, ends_in_five_zeroes)
}

fn part2(secret: &str) -> usize {
    find_hash(secret, ends_in_siz_zeros)
}

fn find_hash(secret: &str, predicate: fn(&[u8]) -> bool) -> usize {
    let base_hasher = Md5::new_with_prefix(secret);
    for i in 0.. {
        let mut hasher = base_hasher.clone();
        hasher.update(i.to_string());
        if predicate(&hasher.finalize()) {
            return i;
        }
    }
    unreachable!()
}

fn ends_in_five_zeroes(hash: &[u8]) -> bool {
    hash[0] | hash[1] | (hash[2] >> 4) == 0
}

fn ends_in_siz_zeros(hash: &[u8]) -> bool {
    hash[0] | hash[1] | hash[2] == 0
}

#[cfg(test)]
mod tests {
    use super::{ends_in_five_zeroes, find_hash};

    #[test]
    fn five_zeros_1() {
        assert_eq!(find_hash("abcdef", ends_in_five_zeroes), 609043);
    }

    #[test]
    fn five_zeros_2() {
        assert_eq!(find_hash("pqrstuv", ends_in_five_zeroes), 1048970);
    }
}
